### PR TITLE
FEATURE: Add setting for group support

### PR DESCRIPTION
**Pull request by oblakeerickson**
Added a site setting so that only admins and any specified groups can
use the needs-love plugin.

### CODE REVIEW

**Comment by tgxworld** [comment]
I'm not sure if this belongs to the `DEV` tag as there is a feature being added here.

File: app/controllers/discourse_needs_love/needs_love_controller.rb [hunk 1]
```diff
@@ -22,5 +23,9 @@ def needs_love
 
       render json: success_json
     end
+
+    def ensure_can_needs_love
+      raise Discourse::InvalidAccess.new unless current_user && current_user.can_needs_love?
```
Comment (tgxworld) [inline app/controllers/discourse_needs_love/needs_love_controller.rb:28]: I think it'll be good if we can test for this.

File: assets/javascripts/discourse-needs-love/initializers/extend-for-needs-love.js.es6 [hunk 2]
```diff
@@ -40,7 +40,10 @@ function registerTopicFooterButtons(api, tagName) {
     classNames: ["needs-love"],
     dependentKeys: ["topic.tags"],
     displayed() {
-      return !disableNeedsLoveButton(this.topic, tagName);
+      return (
+        this.get("currentUser.can_needs_love") &&
+        !disableNeedsLoveButton(this.topic, tagName)
```
Comment (tgxworld) [inline assets/javascripts/discourse-needs-love/initializers/extend-for-needs-love.js.es6:45]: I think this check can be done on the server side and be replaced by something like a `topic.can_needs_love` attribute.

Comment (tgxworld) [reply assets/javascripts/discourse-needs-love/initializers/extend-for-needs-love.js.es6:45]: Personally, I would add an acceptance test to ensure that the button is not displayed but not sure if it is worth it at this point.

Comment (tgxworld) [reply assets/javascripts/discourse-needs-love/initializers/extend-for-needs-love.js.es6:45]: > I think this check can be done on the server side and be replaced by something like a topic.can_needs_love attribute.

Ah if we ship the `can_needs_love` into the current_user serializer, we'll save a query for loading a new topic. I think we can keep this as it is then :)

Comment (tgxworld) [inline assets/javascripts/discourse-needs-love/initializers/extend-for-needs-love.js.es6:45]: Personally, I would add an acceptance test to ensure that the button is not displayed but not sure if it is worth it at this point.

File: plugin.rb [hunk 3]
```diff
@@ -20,4 +20,22 @@
...
+        allowed_groups.present? && groups.exists?(id: allowed_groups) ? :true : :false
+      end
+    end
+
+    @can_needs_love == :true
+  end
+
+  add_to_serializer(:current_user, :can_needs_love) do
+    object.can_needs_love?
+  end
```
Comment (tgxworld) [inline plugin.rb:39]: In order to have test coverage for this and `user.can_needs_love?`, I think we can replace the plugin spec with a current_user serializer spec.

File: spec/requests/discourse_needs_love/needs_love_controller_spec.rb [hunk 4]
```diff
@@ -24,5 +25,30 @@ module DiscourseNeedsLove
...
+
+      before do
+        SiteSetting.needs_love_enabled = true
+        SiteSetting.tagging_enabled = true
+        SiteSetting.needs_love_allowed_groups = "#{group.id}"
+
+        sign_in signed_in_user
+      end
+
+      it 'raises invalid access if not in correct group' do
```
Comment (tgxworld) [inline spec/requests/discourse_needs_love/needs_love_controller_spec.rb:40]:
```suggestion
      it 'raises invalid access if user is not in allowed group' do
```

File: spec/serializers/current_user_serializer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      json = CurrentUserSerializer.new(user, scope: Guardian.new(user), root: false).as_json
+
+      expect(json[:can_needs_love]).to eq(nil)
+    end
+
+    it 'should be false when user can not use needs love' do
+      SiteSetting.needs_love_allowed_groups = "#{group.id}"
+      json = CurrentUserSerializer.new(user, scope: Guardian.new(user), root: false).as_json
+
+      expect(user.can_needs_love?).to eq(false)
```
Comment (tgxworld) [inline spec/serializers/current_user_serializer_spec.rb:26]:
```suggestion
```

I don't think we need this here as we're only concerned about the return value of the serializer.

File: spec/serializers/current_user_serializer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      json = CurrentUserSerializer.new(user, scope: Guardian.new(user), root: false).as_json
+
+      expect(User.find(user.id).can_needs_love?).to eq(true)
+      expect(json[:can_needs_love]).to eq(true)
+    end
+
+    it 'returns true for admins' do
+      json = CurrentUserSerializer.new(admin, scope: Guardian.new(admin), root: false).as_json
+
+      expect(admin.can_needs_love?).to eq(true)
```
Comment (tgxworld) [inline spec/serializers/current_user_serializer_spec.rb:42]:
```suggestion
```

**Comment by oblakeerickson** [comment]
Thank you. Oh I did update the original commit, but I forgot to update the PR with the feature prefix.

**Pull request merged**